src/pages/signin/ValidateCodeForm.js

```diff
@@ -0,0 +1,230 @@
+import withToggleVisibilityView, {toggleVisibilityViewPropTypes} from '../../components/withToggleVisibilityView';
+import canFocusInputOnScreenFocus from '../../libs/canFocusInputOnScreenFocus';
+import * as ErrorUtils from '../../libs/ErrorUtils';
+import {withNetwork} from '../../components/OnyxProvider';
+import networkPropTypes from '../../components/networkPropTypes';
+import OfflineIndicator from '../../components/OfflineIndicator';
+import * as User from '../../libs/actions/User';
+
+const propTypes = {
+    /* Onyx Props */
+
+    /** The details about the account that the user is signing in with */
+    account: PropTypes.shape({
+        /** Whether or not two factor authentication is required */
+        requiresTwoFactorAuth: PropTypes.bool,
+
+        /** Whether or not a sign on form is loading (being submitted) */
+        isLoading: PropTypes.bool,
+    }),
+
+    /** The credentials of the person signing in */
+    credentials: PropTypes.shape({
+        /** The login of the person signing in */
+        login: PropTypes.string,
+    }),
+
+    /** Information about the network */
+    network: networkPropTypes.isRequired,
+
+    ...withLocalizePropTypes,
+    ...toggleVisibilityViewPropTypes,
+};
+
+const defaultProps = {
+    account: {},
+    credentials: {},
+};
+
+class ValidateCodeForm extends React.Component {
+    constructor(props) {
+        super(props);
+        this.validateAndSubmitForm = this.validateAndSubmitForm.bind(this);
+        this.resetValidateCode = this.resetValidateCode.bind(this);
+        this.clearSignInData = this.clearSignInData.bind(this);
+
+        this.state = {
+            formError: false,
+            validateCode: '',
+            twoFactorAuthCode: '',
+        };
+    }
+
+    componentDidMount() {
+        if (!canFocusInputOnScreenFocus() || !this.inputValidateCode || !this.props.isVisible) {
+            return;
+        }
+        this.inputValidateCode.focus();
+    }
+
+    componentDidUpdate(prevProps, prevState) {
+        if (!prevProps.isVisible && this.props.isVisible) {
+            this.inputValidateCode.focus();
+        }
+        if (prevProps.isVisible && !this.props.isVisible && this.state.validateCode) {
+            this.clearValidateCode();
+        }
+        if (!prevProps.account.requiresTwoFactorAuth && this.props.account.requiresTwoFactorAuth) {
+            this.input2FA.focus();
+        }
+        if (prevState.twoFactorAuthCode !== this.state.twoFactorAuthCode && this.state.twoFactorAuthCode.length === CONST.TFA_CODE_LENGTH) {
+            this.validateAndSubmitForm();
+        }
+    }
+
+    /**
+     * Clear Validate Code from the state
+     */
+    clearValidateCode() {
+        this.setState({validateCode: ''}, this.inputValidateCode.clear);
+    }
+
+    /**
+     * Trigger the reset validate code flow and ensure the 2FA input field is reset to avoid it being permanently hidden
+     */
+    resetValidateCode() {
```

Comment:
NAB: I think this should be `resendValidateCode`?

Reply:
Oof yep, good catch 😬